File: HoTT-Report-PDF/src/main/java/de/treichels/hott/report/pdf/ITextInlineImageReplacedElementFactory.java
/**
 * HoTT Transmitter Config Copyright (C) 2013 Oliver Treichel
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
package de.treichels.hott.report.pdf;

import org.w3c.dom.Element;
import org.xhtmlrenderer.extend.ReplacedElement;
import org.xhtmlrenderer.extend.ReplacedElementFactory;
import org.xhtmlrenderer.extend.UserAgentCallback;
import org.xhtmlrenderer.layout.LayoutContext;
import org.xhtmlrenderer.pdf.ITextFSImage;
import org.xhtmlrenderer.pdf.ITextImageElement;
import org.xhtmlrenderer.pdf.ITextReplacedElement;
import org.xhtmlrenderer.render.BlockBox;
import org.xhtmlrenderer.simple.extend.FormSubmissionListener;

import com.itextpdf.text.Image;
import com.itextpdf.text.pdf.codec.Base64;

/**
 * @author Oliver Treichel &lt;oli@treichels.de&gt;
 */
public class ITextInlineImageReplacedElementFactory implements ReplacedElementFactory {
    private static final String PREFIX = "data:image/png;base64,"; //$NON-NLS-1$

    private final ReplacedElementFactory other;

    public ITextInlineImageReplacedElementFactory(final ReplacedElementFactory other) {
        this.other = other;
    }

    @Override
    public ReplacedElement createReplacedElement(final LayoutContext c, final BlockBox box, final UserAgentCallback uac, final int cssWidth,
            final int cssHeight) {
        final Element elem = box.getElement();

        // check if we have an inline png image
        if (!(elem != null && elem.getNodeName().equals("img") && elem.hasAttribute("src") && elem.getAttribute("src").startsWith(PREFIX))) return null; // other.createReplacedElement(c,
                                                                                                                                                         // box,
                                                                                                                                                         // uac,
                                                                                                                                                         // cssWidth,
        // cssHeight);

        int width = 0;
        int height = 0;

        if (cssWidth > 0) width = cssWidth;
        if (cssHeight > 0) height = cssHeight;

        if (elem.hasAttribute("width")) width = Integer.parseInt(elem.getAttribute("width")); //$NON-NLS-1$

        if (elem.hasAttribute("height")) height = Integer.parseInt(elem.getAttribute("height")); //$NON-NLS-1$

        final String inlineData = elem.getAttribute("src").substring(PREFIX.length()); // strip //$NON-NLS-1$
                                                                                       // leading
                                                                                       // "data:image/png;base64,"

        try {
            final Image image = Image.getInstance(Base64.decode(inlineData));

            if (width == 0) width = (int) image.getWidth();

            if (height == 0) height = (int) image.getHeight();

            image.scaleAbsolute(width * 16, height * 16);

            final ITextFSImage fsImage = new ITextFSImage(image);
            final ITextReplacedElement element = new ITextImageElement(fsImage);
            return element;
        } catch (final Exception e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public void remove(final Element e) {
        other.remove(e);
    }

    @Override
    public void reset() {
        other.reset();
    }

    @Override
    public void setFormSubmissionListener(final FormSubmissionListener listener) {
        other.setFormSubmissionListener(listener);
    }
}

File: HoTT-Report-PDF/src/main/kotlin/de/treichels/hott/report/pdf/ITextInlineImageReplacedElementFactory.kt
/*
  HoTT Transmitter Config Copyright (C) 2013 Oliver Treichel

  This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software
  Foundation, either version 3 of the License, or (at your option) any later version.

  This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
  A PARTICULAR PURPOSE. See the GNU General Public License for more details.

  You should have received a copy of the GNU General Public License along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
package de.treichels.hott.report.pdf

import com.itextpdf.text.Image
import com.itextpdf.text.pdf.codec.Base64
import org.w3c.dom.Element
import org.xhtmlrenderer.extend.ReplacedElement
import org.xhtmlrenderer.extend.ReplacedElementFactory
import org.xhtmlrenderer.extend.UserAgentCallback
import org.xhtmlrenderer.layout.LayoutContext
import org.xhtmlrenderer.pdf.ITextFSImage
import org.xhtmlrenderer.pdf.ITextImageElement
import org.xhtmlrenderer.render.BlockBox
import org.xhtmlrenderer.simple.extend.FormSubmissionListener

/**
 * @author Oliver Treichel &lt;oli@treichels.de&gt;
 */
internal class ITextInlineImageReplacedElementFactory(private val other: ReplacedElementFactory) : ReplacedElementFactory {
    override fun createReplacedElement(c: LayoutContext, box: BlockBox, uac: UserAgentCallback, cssWidth: Int, cssHeight: Int): ReplacedElement? {
        val elem = box.element

        // check if we have an inline png image
        if (!(elem != null && elem.nodeName == "img" && elem.hasAttribute("src") && elem.getAttribute("src").startsWith(PREFIX))) return null

        var width = 0
        var height = 0

        if (cssWidth > 0) width = cssWidth
        if (cssHeight > 0) height = cssHeight

        if (elem.hasAttribute("width")) width = Integer.parseInt(elem.getAttribute("width"))
        if (elem.hasAttribute("height")) height = Integer.parseInt(elem.getAttribute("height"))

        // strip leading "data:image/png;base64,"
        val inlineData = elem.getAttribute("src").substring(PREFIX.length)

        return try {
            val image = Image.getInstance(Base64.decode(inlineData))
            if (width == 0) width = image.width.toInt()
            if (height == 0) height = image.height.toInt()

            image.scaleAbsolute((width * 16).toFloat(), (height * 16).toFloat())

            val fsImage = ITextFSImage(image)
            ITextImageElement(fsImage)
        } catch (e: Exception) {
            throw RuntimeException(e)
        }
    }

    override fun remove(e: Element) {
        other.remove(e)
    }

    override fun reset() {
        other.reset()
    }

    override fun setFormSubmissionListener(listener: FormSubmissionListener) {
        other.setFormSubmissionListener(listener)
    }

    companion object {
        private const val PREFIX = "data:image/png;base64,"
    }
}